src/Event/EventStack.php

```diff
@@ -0,0 +1,136 @@
+<?php
+/**
+ * CakePHP(tm) : Rapid Development Framework (http://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * @link          http://cakephp.org CakePHP(tm) Project
+ * @since         3.3.0
+ * @license       http://www.opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Event;
+
+use Cake\Event\Event;
+
+/**
+ * The Event Stack
+ */
+class EventStack implements \ArrayAccess, \Countable
+{
+
+    /**
+     * Events list
+     *
+     * @var array
+     */
+    protected $_events = [];
+
+    /**
+     * Empties the stack of dispatched events.
+     *
+     * @return void
+     */
+    public function flush()
+    {
+        $this->_events = [];
+    }
+
+    /**
+     * Adds an event to the stack when stacking is enabled.
+     *
+     * @param \Cake\Event\Event $event An event to stack.
+     * @return void
+     */
+    public function add(Event $event)
+    {
+        $this->_events[] = $event;
```

Comment:
This isn't _really_ a 'stack'. A stack usually prepends into the underlying list.